flight/mcu/lib/teensyShared/ActiveControl.cpp

```diff
@@ -1,95 +1,123 @@
+
+    float min = abs(u - prev_u)/dt;
+
+    if (du_max < min) {
+        min = du_max;
     }
 
-    // low pass filter
-    if (l1_cmd < 0.002794) {
-        l1_cmd = 0;
+    float sign = 1;
+
+    if (u - prev_u < 0) {
+        sign = -1;
     }
-    if (l2_cmd < 0.002794) {
-        l2_cmd = 0;
+    u = u + sign*min*dt;
+    prev_u = u;
+
+
+    //Set flap extension limits
+    if (u < min_extension) {
+        pointer_struct->sensorDataPointer->flap_data.l1=u;
+        u = min_extension;
+    } else if (u > max_extension) {
+        pointer_struct->sensorDataPointer->flap_data.l1=0;
+        u = max_extension;
     }
 
+    // std::cout << "Controller Flap Extension: " << u << std::endl;
+
+    
     if (ActiveControl_ON()) {
-        activeControlServos.servoActuation(l1_cmd, l2_cmd);
+        activeControlServos.servoActuation(u);
+        
     } else {
-        activeControlServos.servoActuation(0, 0);
+        if (pointer_struct->sensorDataPointer->rocketState_data.rocketState==STATE_APOGEE) {
+            activeControlServos.servoActuation(0);
+        } else {
+            activeControlServos.servoActuation(15);
+        }
     }
 
-    e_prev = e;
-    l1_prev = l1_cmd;
-    l2_prev = l2_cmd;
 }
 
-bool ActiveControl::ActiveControl_ON() {
-    bool active_control_on = false;
+bool Controller::ActiveControl_ON() {
+
+    // For some reason, this only sees IDLE through launch.
+
+    bool active_control_on = true;
     switch (*current_state) {
         case STATE_INIT:
+            // std::cout << "INIT" << std::endl;
             active_control_on = false;
             break;
         case STATE_IDLE:
+            // std::cout << "IDLE" << std::endl;
             active_control_on = false;
             break;
         case STATE_LAUNCH_DETECT:
+            // std::cout << "Launch Detect" << std::endl;
             active_control_on = false;
             break;
         case STATE_BOOST:
+            // std::cout << "BOOST" << std::endl;
             active_control_on = false;
             break;
         case STATE_COAST:
-            active_control_on = true;
+            // std::cout << "COAST" << std::endl;
+            // This adds a delay to the 
+            if (*ac_coast_timer > coast_ac_delay_thresh) {
+                active_control_on = true;
+            }            
             break;
         case STATE_APOGEE_DETECT:
+            // std::cout << "Apogee" << std::endl;
             active_control_on = false;
             break;
         default:
             active_control_on = false;
             break;
     }
     return active_control_on;
+}
+
+void Controller::setLaunchPadElevation(){
+    float sum = 0;
+    for(int i = 0; i < 30; i++){
+        chMtxLock(dataMutex_barometer_);
+        sum += *b_alt;
+        chMtxUnlock(dataMutex_barometer_);
+        chThdSleepMilliseconds(100);
+    }
+    launch_pad_alt = sum / 30;
+    apogee_des_msl = apogee_des_agl + launch_pad_alt;
```

Comment:
@parthshrotri This is what I also thought. You seemed to insist that calculating it dynamically was a better choice. Do you want to stick with this or just hard code it for each launch since we can get an exact number for altitude above sea level just by looking it up. Barometers always have a reasonably large uncertainty

Reply:
Yes this does need to be calculated dynamically because the altitude reference frame is barometric, ie. with air pressure which varies from day to day. Think setting the barometric pressure in the altimeter of a plane as a part of preflight checks. If a barometer reads 101325Pa, it will say that it is at 0ft Mean Sea Level even if it is at a different altitude. This can vary significantly hour to hour and day to day which is why we came up with this solution

Reply:
I will write a comment in this and the Kalman filter files once I figure out how to